xcm/xcm-simulator/example/src/lib.rs

```diff
@@ -202,14 +216,165 @@ mod tests {
+			let message = WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset { assets: vec![All], dest: Parachain(2).into() },
+				],
+			};
+			// Send withdraw and deposit
+			assert_ok!(ParachainPalletXcm::send_xcm(Null, X1(Parent), message.clone(),));
+		});
+
+		Relay::execute_with(|| {
+			let amount = 10;
+			let para_account_a: relay_chain::AccountId = ParaId::from(1).into_account();
+			assert_eq!(
+				relay_chain::Balances::free_balance(para_account_a),
+				INITIAL_BALANCE - amount
+			);
+			let para_account_b: relay_chain::AccountId = ParaId::from(2).into_account();
+			assert_eq!(relay_chain::Balances::free_balance(para_account_b), 10);
+		});
+	}
+
+	/// Scenario:
+	/// A parachain wants to be notified that a transfer worked correctly.
+	/// It sends a `QueryHolding` after the deposit to get notified on success.
+	///
+	/// Asserts that the balances are updated correctly and the expected XCM is sent.
+	#[test]
+	fn query_holding() {
+		MockNet::reset();
+
+		// First send a message which fails on the relay chain
+		ParaA::execute_with(|| {
+			let amount = 10;
+			let weight = 3 * relay_chain::BaseXcmWeight::get();
+			//let para_account_b: relay_chain::AccountId = ParaId::from(2).into_account();
+			let query_id = 1234;
+			let message = WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: OnlyChild.into(), // invalid destination
+					},
+					// is not triggered because the deposit fails
+					Order::QueryHolding { query_id, dest: Parachain(2).into(), assets: vec![All] },
+				],
+			};
+			// Send withdraw and deposit with query holding
+			assert_ok!(ParachainPalletXcm::send_xcm(Null, X1(Parent), message.clone(),));
+		});
+
+		// Check that no transfer was executed and no response message was sent
+		Relay::execute_with(|| {
+			let amount = 10;
+			let para_account_a: relay_chain::AccountId = ParaId::from(1).into_account();
+			// withdraw did execute
+			assert_eq!(
+				relay_chain::Balances::free_balance(para_account_a),
+				INITIAL_BALANCE - amount
+			);
+			// but deposit did not execute
+			let para_account_b: relay_chain::AccountId = ParaId::from(2).into_account();
+			assert_eq!(relay_chain::Balances::free_balance(para_account_b), 0);
+		});
+
+		// Now send a message which fully succeeds on the relay chain
+		ParaA::execute_with(|| {
+			let amount = 10;
+			let weight = 3 * relay_chain::BaseXcmWeight::get();
+			let query_id = 1234;
+			let message = WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: Parachain(2).into(), // valid destination
+					},
+					Order::QueryHolding { query_id, dest: Parachain(2).into(), assets: vec![All] },
```

Comment:
The purpose of this test is to check that if all the previous messages execute successfully, a `Order::QueryResponse` will be sent to this `dest`.

The `dest` should be `Parachain(1)` instead of `Parachain(2)`, but when I make the change shown below, there is an `BorrowMut` error because it triggers `execute_with` inside of another `execute_with`
```suggestion
					Order::QueryHolding { query_id, dest: Parachain(1).into(), assets: vec![All] },
```
The idea of a solution is to make message execution less automatic so that SEND, RECEIVE, and PROCESS are separate steps.

Reply:
@shaunxw This is the issue I mentioned in chat.

Reply:
As discussed in the call, the less automatic approach sounds useful in some scenarios.

I'm just thinking that, as in most unit tests here and in orml-xtokens, we're expecting message passing in an automatic way. Probably we can make it automatic by default, while providing a flag in the mock network decl macro to allow users to make it manual. This could keep APIs still simple to use.

Reply:
#3552